sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+and define batched loaders for loading tensors. 
+
+
+### Options and Builder Classes
+
+To enable methods and constructors to use optional arguments with defaults,
+`Options` classes are defined (most notably for `Layer` construction,
+and `compile` and `fit` configuration for models). The structure for an `Options` class
+of object `KerasObject` is standardized as follows; if changes can be made to this structure
+so that it is cleaner and easier to work with, please comment!
+
+```java
+class KerasObject {
+
+    static {
+        Arg1 arg1;
+        Arg2 arg2;
+        
+        // Create KerasObject from Options class
+        KerasObject kerasObject =
+            new KerasObject(KerasObject.Options.builder()           
+                .setArg1(arg1)
+                .setArg2(arg2)
+                .build());
+    }
+
+    public KerasObject(Options options) {
+        // construct KerasObject
+    }    
+
+    public static class Options() {
+        Arg1 arg1;
+        Arg2 arg2
+
+        private void defaults() {
+            return new Builder()
+                .setArg1(DEFAULT_ARG1)
+                .setArg2(DEFAULT_ARG2)
+                .build()
+        }
+
+        public Builder builder() {
+            return defaults();
+        }
+
+        public static class Builder() {
+            Options options;
+
+            public Builder() {
+                options = new Options()
+            }
+
+            public Builder setArg1(Arg1 arg1) {
+                options.arg1 = arg1;
+                return this;
+            }
+            public void setArg2(Arg2 arg2) {
+                options.arg2 = arg2;
+                return this;
+            }
+
+            public Options build() {
+                return options;
+            }
+        }
+    }
+}
+```
+
+### Comparing Java API to Python
+#### Original Python MNIST
+```python
+import tensorflow as tf
+
+model = tf.keras.models.Sequential([
+  tf.keras.layers.Flatten(input_shape=(28, 28)),
+  tf.keras.layers.Dense(128, activation='relu', kernel_initializer="random_normal", bias_initializer="zeros"),
+  tf.keras.layers.Dense(10, activation='softmax', kernel_initializer="random_normal", bias_initializer="zeros")
+])
+
+model.compile(optimizer='sgd', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
+
+(X_train, y_train), (X_val, y_val) = tf.keras.datasets.load_mnist()
+model.fit(X_train, y_train, val_data=(X_val, y_val), epochs=10, batch_size=100)
+```
+
+#### MNIST using Java API
+
+```java
+public class MNISTKeras {
+    private static Model<Float> model;
+
+    static {
+        model = Sequential.of(
+            Float.class, // dtype which will be passed through layer.build
+            Layers.input(28, 28),
+            Layers.flatten(),
+            Layers.dense(128, Activations.softmax, Initializers.randomNormal, Initializers.zeros),
+            Layers.dense(10, Activations.softmax, Initializers.randomNormal, Initializers.zeros)
+        );
+    }
```

Comment:
This looks very clean! It would be nice if we can help achieving a similar result when using a user-defined layers.